### PR TITLE
[Spark-23240][python] Better error message when extraneous data in pyspark.daemon's stdout

### PR DESCRIPTION
**Pull request by bersprockets**
## What changes were proposed in this pull request?

Print more helpful message when daemon module's stdout is empty or contains a bad port number. 

## How was this patch tested?

Manually recreated the environmental issues that caused the mysterious exceptions at one site. Tested that the expected messages are logged.

Also, ran all scala unit tests.

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by vanzin** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
Ah, OK. fixing the error message is fine as a separate improvement.

**Comment by HyukjinKwon** [comment]
retest this please

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 1]
```diff
@@ -191,7 +191,20 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
         daemon = pb.start()
 
         val in = new DataInputStream(daemon.getInputStream)
-        daemonPort = in.readInt()
+        try {
+          daemonPort = in.readInt()
+        } catch {
+          case exc: EOFException =>
+            throw new IOException(s"No port number in $daemonModule's stdout")
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Shall we throw `SparkException(s"No port number in $daemonModule's stdout", eof)`?

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 2]
```diff
@@ -191,7 +191,20 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
...
+        }
+
+        // test that the returned port number is within a valid range.
+        // note: this does not cover the case where the port number
+        // is arbitrary data but is also coincidentally within range
+        if (daemonPort < 1 || daemonPort > 0xffff) {
+          throw new IOException(s"Bad port number in $daemonModule's stdout: " +
+            f"0x$daemonPort%08x")
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: `SparkException` too.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 3]
```diff
@@ -191,7 +191,20 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
         daemon = pb.start()
 
         val in = new DataInputStream(daemon.getInputStream)
-        daemonPort = in.readInt()
+        try {
+          daemonPort = in.readInt()
+        } catch {
+          case exc: EOFException =>
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: nit: `exc` -> `_`.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 4]
```diff
@@ -191,7 +191,20 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
...
+            throw new IOException(s"No port number in $daemonModule's stdout")
+        }
+
+        // test that the returned port number is within a valid range.
+        // note: this does not cover the case where the port number
+        // is arbitrary data but is also coincidentally within range
+        if (daemonPort < 1 || daemonPort > 0xffff) {
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Why don't we just copy the condition in `java.net.InetSocketAddress.checkPort`: `if (port < 0 || port > 0xFFFF)`?

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala (hunk 2 above)
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Shall we also show the actual number?

**Comment by HyukjinKwon** [comment]
LGTM except two nits and one question.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala (hunk 2 above)
Comment (squito) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: just a thought:

 this error message won't mean much to the typical user.  Would it be sensible to tell the user exactly what python command to run themselves to figure out the problem?  Eg. "unexpected stdout from /foo/bar/some/path/to/python -m /path/to/daemon.py".  That's what would help with that sitecustomization.py case.  Or not useful in general?

Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Yes, that makes sense. I might not be able to get it as clear as the exact path to the file, since the PythonWorkerFactory sets a PYTHONPATH environmental variable and then lets python itself figure out where on those paths the module actually lives. But I could tell the user how the PYTHONPATH was set up (in a generic sense, without using any shell's syntax) and then how the python command was subsequently run.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala (hunk 4 above)
Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Port 0 has special meaning. A program passes port 0 when it wants the system to choose an unused port on the program's behalf. So, the daemon should not return 0.

It's valid to pass port 0 to InetSocketAddress, since you might be asking for the system to assign a port for you.

However, following my own logic, the code in my pull request really should be checking for the range 49152-65535 (ephemeral range) instead of 1-65535, but I didn't have the nerve to make it that restrictive.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala (hunk 3 above)
Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: I see, exc not used.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala (hunk 4 above)
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: I am saying we can just safely move the same check and fail fast, which is simple and theoretically safe.

Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Sorry @HyukjinKwon , I didn't quite get the last point. Could you rephrase?

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: I mean, I left my sign-off because what we do is basically move the _same_ check (`java.net.InetSocketAddress.checkPort`) ahead and another one is simply to wraps an exception, `EOFException`. I think we are here safe in theory.

I got your point of reserved ports and now the condition became narrower. I should check other things like which error it produces before in this case and if the current error message is nicer. Also, this seems not completely addressing the concerns about it.

I was wondering if this is worth doing these stuff. If you strongly prefer this, I won't stay against but may request few more investigations.

Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Oh, I see. Let me address the two parts of the comment separately:

First part: What's the point of throwing an exception for a bad port number when the original handling did that already?

The original handling was:

<pre>
 java.lang.IllegalArgumentException: port out of range:1315905645
        at java.net.InetSocketAddress.checkPort(InetSocketAddress.java:143)
</pre>

This error occurred in a different function than the one that obtained the port number. So you had to track down the source of the port number. This actually added an extra step to the original debugging for the sitecustomize.py issue.

The proposed handling is:

<pre>
java.io.IOException: Bad port number in pyspark.daemon's stdout: 0x4e6f206d
        at org.apache.spark.api.python.PythonWorkerFactory.startDaemon(PythonWorkerFactory.scala:205)
</pre>

Here we're not saying "somehow we got a bad port number", but that a particular python module (the name of which is displayed, since the name of that module is configurable) has returned bad data.

Also, the check is at the point at which the port number is obtained, not waiting until sometime later in another function where the port number is used (and where possibly something else has changed the port number, which is not likely, but you would need to check that).

Perhaps the message could be better, e.g.,:

<pre>
Bad data in pyspark.daemon's output.
Expected valid port number, got 0x4e6f206d.
PYTHONPATH set to '/Users/brobbins/github/spark_fork/python/lib/pyspark.zip:/Users/brobbins/github/spark_fork/python/lib/py4j-0.10.6-src.zip:/Users/brobbins/github/spark_fork/assembly/target/scala-2.11/jars/spark-core_2.11-2.4.0-SNAPSHOT.jar:/Users/brobbins/github/spark_fork/python/lib/py4j-0.10.6-src.zip:/Users/brobbins/github/spark_fork/python/:'
Command to run python module was 'python -m pyspark.daemon'
Check whether you have a sitecustomize.py module that may be printing output to stdout.
</pre>

Second part: Why don't we fix this?

That's reasonable. A couple of points:

- The name of the python daemon module is now configurable so that the module can be wrapped with customizations. It appears that this is only in the main branch and not even released on Spark 2.3, so it might be safe to change daemon.py (and potentially its existing wrappers) to return the port number in a different way.
- I don't have a good feel for how often sitecustomize.py is used, so not sure of the relative value of some mild hacking up of this code vs. just letting the user know what happened.

Comment (squito) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: yeah I could go either way on this.  Personally i think just the enhanced error message you have above would be useful, without going through the added trouble of using another approach.  I'll defer to your opinion @HyukjinKwon

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Thanks, @squito. My only worry is, what is the error message like befer / after this change when `daemonPort` happens to be 0 because it was filled in pyspark.daemon's stdout unexpectedly by the sitecustomize or usercustomize whatever?

I assume the error message becomes different as you described above because `java.net.InetSocketAddress.checkPort` passes `0` case but your proposed change also catches `0` case too for the reason you described above.

This case `0` is different because the previous codes probably try to attempt a connection whether it's a valid port or not, after passing `java.net.InetSocketAddress.checkPort`, if I didn't misunderstand.

I want to be very sure on this. Is it guranteed to be always failed without any hole and the error message is always better in this case? How about other ports reserved or "well-known" ? Should we really take this into account here?

Please let me know if I missed something.

Comment (squito) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: ah I see, I think you are worried about something other than what bruce and I thought.  Your concern is that we might throw an exception for some values that are actually perfectly legitimate.  Port 0 being special is a pretty standard thing -- its mentioned in the constructor for ServerSocket: https://docs.oracle.com/javase/7/docs/api/java/net/ServerSocket.html#ServerSocket%28int%29

which implies that you shouldn't ever open a Socket on port 0, though I don't see that officially documented.  At least on my laptop, I get different errors if I try to connect to port 0, vs. just connecting to a bogus port:

```scala
scala> val s2 = new Socket("localhost", 1234)
java.net.ConnectException: Connection refused
  at java.net.PlainSocketImpl.socketConnect(Native Method)
  at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
  at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
  at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
  at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
  at java.net.Socket.connect(Socket.java:589)
  at java.net.Socket.connect(Socket.java:538)
  at java.net.Socket.<init>(Socket.java:434)
  at java.net.Socket.<init>(Socket.java:211)
  ... 29 elided

scala> val s3 = new Socket("localhost", 0)
java.net.NoRouteToHostException: Can't assign requested address
  at java.net.PlainSocketImpl.socketConnect(Native Method)
  at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
  at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
  at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
  at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
  at java.net.Socket.connect(Socket.java:589)
  at java.net.Socket.connect(Socket.java:538)
  at java.net.Socket.<init>(Socket.java:434)
  at java.net.Socket.<init>(Socket.java:211)
  ... 29 elided
```

so I think its pretty safe to say that daemon.py (or whatever) shouldn't be passing back `0` as the port to bind to.

Still -- it is *clearly* safer to instead have the port written to some other file, or (another) socket, so that you we wouldn't have to worry about the details of this error handling.

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Ah, OK. Thanks for clarification. Maybe, I was caring too much about it. Thanks all for bearing with me. I am fine as is.

Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: > Still -- it is clearly safer to instead have the port written to some other file, or (another) socket, so that you we wouldn't have to worry about the details of this error handling.

Let's go that route. I will have a PR in the coming days.

**Comment by HyukjinKwon** [comment]
So, I think I am fine. I believe it's already a corner case and I think we don't have to put a lot of efforts on this ... to be honest. I only wanted to double check if there is a slight step back in the case of `0`.

@vanzin, @squito and @bersprockets, so, are we fine with that error message for now here?

I was thinking we could just go ahead but I believe it's nicer if I can double check / make sure before going and merging this one.

Do you guys have some more information to add in the error message?

**Comment by bersprockets** [comment]
Hi @HyukjinKwon 

>Do you guys have some more information to add in the error message?

If we go with this solution, I would like to improve the message before the PR is merged.

**Comment by bersprockets** [comment]
@HyukjinKwon @squito 

Before merging this, allow me make an alternative PR that actually bypasses the problem (rather than simply reporting on the problem). If it seems like the solution adds to much complexity, we can reject the alternate PR and we can go with this one instead. Does that make sense?

**Comment by HyukjinKwon** [comment]
Yup, makes sense.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 5]
```diff
@@ -191,7 +192,25 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
...
+          val exceptionMessage = f"""
+               |Bad data in  $daemonModule's standard output.
+               |Expected valid port number, got 0x$daemonPort%08x.
+               |PYTHONPATH set to '$pythonPath'
+               |Python command is '${command.asScala.mkString(" ")}'
+               |Check if you have a sitecustomize.py module in your python installation."""
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: This sounds like there's only one case - `sitecustomize`. I believe there are many possibility, for example, custom python executable, `usercustomize` and etc. Can we just say few words to implies `sitecustomize` is one potential case?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Will take a final look for this if we are all fine with this for now.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by squito** [comment]
lgtm
@bersprockets you mentioned wanting to try the other route as well -- whats the status on that?  shoudl we still wait on this one?

**Comment by bersprockets** [comment]
@squito 

>you mentioned wanting to try the other route as well

That was this PR: https://github.com/apache/spark/pull/20519

@HyukjinKwon would prefer to go with the error message for now and wait and see if the issue comes up again.

**Comment by squito** [comment]
ah got it.  sounds good to me, I will defer to @HyukjinKwon 's judgement.  I think this change looks fine

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 6]
```diff
@@ -191,7 +192,26 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
...
+        // note: this does not cover the case where the port number
+        // is arbitrary data but is also coincidentally within range
+        if (daemonPort < 1 || daemonPort > 0xffff) {
+          val exceptionMessage = f"""
+               |Bad data in  $daemonModule's standard output.
+               |Expected valid port number, got 0x$daemonPort%08x.
+               |PYTHONPATH set to '$pythonPath'
+               |Python command is '${command.asScala.mkString(" ")}'
+               |One possibility is a sitecustomize.py module in your python installation
+               |that is printing to stdout"""
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Shall we keep the format same as:

https://github.com/apache/spark/blob/b63abee881f2b4379f375500d51fdef706d6d512/core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala#L232-L235

?

the current message looks:

```
...
Caused by: java.io.IOException:
Bad data in  pyspark.daemon's standard output.
Expected valid port number, got 0x4920616d.
PYTHONPATH set to '/.../spark/python/lib/pyspark.zip:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/assembly/target/scala-2.11/jars/spark-core_2.11-2.4.0-SNAPSHOT.jar:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/python/:'
Python command is 'python -m pyspark.daemon'
Check if you have a sitecustomize.py module in your python installation.
...
```

I made a suggestion while verifying this PR:


```
...
Error from bad data in pyspark.daemon's standard output. Invalid port number:
  1633771786 (0x6161610a)
Python command to execute the daemon was:
  python -m pyspark.daemon

One possibility is a sitecustomize module printing some data to the standard output.
This module 'sitecustomize.py' can be located in your Python path:
  /.../spark/python/lib/pyspark.zip:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/assembly/target/scala-2.11/jars/spark-core_2.11-2.4.0-SNAPSHOT.jar:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/python/:
...
```

Here is the diff I used. I also did some insane nitpicks here as well.

```diff
diff --git a/core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala b/core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala
index 5790c050a7f..b44aa6064bb 100644
--- a/core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala
+++ b/core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala
@@ -196,20 +196,24 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
           daemonPort = in.readInt()
         } catch {
           case _: EOFException =>
-            throw new IOException(s"No port number in $daemonModule's stdout")
+            throw new SparkException(s"No port number in $daemonModule's standard output.")
         }

-        // test that the returned port number is within a valid range.
-        // note: this does not cover the case where the port number
-        // is arbitrary data but is also coincidentally within range
+        // Check if the returned port number is within a valid range.
+        // Note: this does not cover the case where the port number is arbitrary data but is
+        // also coincidentally within range.
         if (daemonPort < 1 || daemonPort > 0xffff) {
           val exceptionMessage = f"""
-               |Bad data in  $daemonModule's standard output.
-               |Expected valid port number, got 0x$daemonPort%08x.
-               |PYTHONPATH set to '$pythonPath'
-               |Python command is '${command.asScala.mkString(" ")}'
-               |Check if you have a sitecustomize.py module in your python installation."""
-          throw new IOException(exceptionMessage.stripMargin)
+             |Error from bad data in $daemonModule's standard output. Invalid port number:
+             |  $daemonPort (0x$daemonPort%08x)
+             |Python command to execute the daemon was:
+             |  ${command.asScala.mkString(" ")}
+             |
+             |One possibility is a sitecustomize module printing some data to the standard output.
+             |This module 'sitecustomize.py' can be located in your Python path:
+             |  $pythonPath
+             |"""
+          throw new SparkException(exceptionMessage.stripMargin)
         }

         // Redirect daemon stdout and stderr
```

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 7]
```diff
@@ -191,7 +192,26 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
...
+          throw new IOException(exceptionMessage.stripMargin)
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: I think this rather should be `SparkException`.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 8]
```diff
@@ -191,7 +192,26 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
         daemon = pb.start()
 
         val in = new DataInputStream(daemon.getInputStream)
-        daemonPort = in.readInt()
+        try {
+          daemonPort = in.readInt()
+        } catch {
+          case _: EOFException =>
+            throw new IOException(s"No port number in $daemonModule's stdout")
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Here too, `SparkException`.

**Comment by HyukjinKwon** [comment]
Thank you @squito  .. LGTM otherwise.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 9]
```diff
@@ -17,7 +17,7 @@
 
 package org.apache.spark.api.python
 
-import java.io.{DataInputStream, DataOutputStream, InputStream, OutputStreamWriter}
+import java.io._
```
Comment (HyukjinKwon) [inline core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: BTW, don't forget to get rid of this import change too. This import looks needed due to `IOException`.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala (hunk 6 above)
Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Nice, except this one thing:

>This module 'sitecustomize.py' can be located in your Python path:
  /.../spark/python/lib/pyspark.zip:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/assembly/target/scala-2.11/jars/spark-core_2.11-2.4.0-SNAPSHOT.jar:/.../spark/python/lib/py4j-0.10.6-src.zip:/.../spark/python/:

I display the path because maybe you accidentally configured the path to have some old .zip or other incompatible versions of expected python modules on your path.

sitecustomize.py might not be in your path, although it could be. I found a machine that had it here: /usr/lib/python2.7/sitecustomize.py. Another, I found it here: /usr/lib64/python2.7/sitecustomize.py

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: Oh, I see. I meant it might be located. please fix few my bad words :). One thing I was worried of is, it might print the Python path twice if it contained stderr (it can be wrapped in L232 -L235).

I think in most cases the Python paths will be printed out actually because it should usually throw an exception.

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: But I think there's no need to put some additional logics here to deal with it tho. Let's just fix the error message to be in the similar format.

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala (hunk 9 above)
Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: To save on jenkins resources :), I want to verify that the latest version of the message is OK:

<pre>
org.apache.spark.SparkException:
Bad data in pyspark.daemon's standard output. Invalid port number:
  1315905645 (0x4e6f206d)
Python command to execute the daemon was:
  python -m pyspark.daemon
Check that you don't have any unexpected modules or libraries in
your PYTHONPATH:
  /.../github/spark_fork/python/lib/pyspark.zip:/.../github/spark_fork/python/lib/py4j-0.10.6-src.zip:/.../github/spark_fork/assembly/target/scala-2.11/jars/spark-core_2.11-2.4.0-SNAPSHOT.jar:/.../github/spark_fork/python/lib/py4j-0.10.6-src.zip:/.../github/spark_fork/python/:
Also, check if you have a sitecustomize.py module in your python path,
or in your python installation, that is printing to standard output

</pre>

Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: 👍

File: core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala [hunk 10]
```diff
@@ -191,7 +192,29 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
...
+          case _: EOFException =>
+            throw new SparkException(s"No port number in $daemonModule's stdout")
+        }
+
+        // test that the returned port number is within a valid range.
+        // note: this does not cover the case where the port number
+        // is arbitrary data but is also coincidentally within range
+        if (daemonPort < 1 || daemonPort > 0xffff) {
+          val exceptionMessage = f"""
+               |Bad data in $daemonModule's standard output. Invalid port number:
```
[the remark this answers is not in the document]
Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: >Oh, I believe double spaces were actually a style nit just to be consistent.

Oops, did I miss something?

[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: I meant ... 

```scala
val exceptionMessage = f"""
  |...
  |...
...
```

[the remark this answers is not in the document]
Comment (bersprockets) [reply core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala]: OK, I will change that.

**Comment by HyukjinKwon** [comment]
LGTM!

**Comment by bersprockets** [comment]
The test that finished last succeeded, but the one that started last had a spurious error. Can I get a retest?

**Comment by HyukjinKwon** [comment]
Sure, you can. retest this please

**Comment by bersprockets** [comment]
@squito We made a few adjustments since your "lgtm". Do you want to take a quick look? @HyukjinKwon also gave his "lgtm" after the adjustments.

**Comment by squito** [comment]
still lgtm, thanks

**Comment by HyukjinKwon** [comment]
Thanks @squito. Will merge this one in few days.

**Comment by HyukjinKwon** [comment]
Merged to master.
